include/os/linux/kernel/linux/simd_x86.h

```diff
@@ -135,7 +135,10 @@
  */
 #if defined(HAVE_KERNEL_FPU_INTERNAL)
 
+#include <sys/zfs_context.h>
```

Comment:
This shouldn't be needed and can be removed as long as we only rely of the kernel slab interfaces.

Reply:
thought so as well at first, but unfortunately:
```
In file included from /home/fgruenbichler/ext/zfs-upstream/include/os/linux/kernel/linux/simd.h:29,
                 from /home/fgruenbichler/ext/zfs-upstream/include/os/linux/spl/sys/simd.h:29,
                 from /home/fgruenbichler/ext/zfs-upstream/module/icp/algs/modes/gcm.c:31:
/home/fgruenbichler/ext/zfs-upstream/include/os/linux/kernel/linux/simd_x86.h:142:2: error: #error "Kernel slab allocator must be used"
 #error "Kernel slab allocator must be used"
```

my first version just used the regular kernel interface ;) it would be nice if that would still be accessible somehow even with zfs_context.h (transitively) included.. move the kfpu implementation to a module/zcommon and undef the spl wrappers in the .c file, similar to what spl-kmem-cache.c does?

Reply:
or undef them, and then re-define them at the end of simd_x86.h? just pushed a version of this - it's not very pretty, but way less work than moving big chunks of code around ;)

Reply:
You're right, I mistakenly compiled the non-internal version when testing this locally.  Whoops!

> it's not very pretty,

Agreed.  Given the complications involved in bypassing the SPL I feel like we should avoid this approach.  That leaves us with two pretty reasonable options from my perspective.

* Revert to your original approach and use the spl kmem cache interfaces.  That gets rid of the worst of the ugliness, but it does muddle the laying somewhat since this Linux specific code is now depending on spl functionality.  Additionally, there's the numa node issue.  I pushed a new branch with a few fixes on top of your original change for comment.

    https://github.com/behlendorf/zfs/commits/9608-fix-simd-alignment-v3

* Or, we don't use a kmem cache to satisfy the alignment restriction.  Instead we use `alloc_pages_node()` and `free_pages()` to allocate full pages.  This has the advantage of keeping the layering clear and we can request the preferred numa node.  The downside being this may take more memory compared to the Linux kmem cache, but that would really depend on how/if the allocations are packed.  It may not end us using more memory.

    https://github.com/behlendorf/zfs/commits/9608-fix-simd-alignment-v2

Looking at both approaches I'm leaning towards the simpler `alloc_pages_node()` approach.  But what are your thoughts?

One last thing, I've only lightly tested the two referenced branches on a stock Fedora 5.3.13-300.fc31.x86_64 kernel.  If you could verify the `alloc_pages_node()` does resolve the issue in your test environment that would be great.

Reply:
:thumbsup: `alloc_pages_node()`

Reply:
-v2 looks good to me and passes a simple smoke-test as well.

since it's limited to x86, we also don't have to worry about bigger than 4k pages wasting space. the union in question is one 4k page big, but it is 75% padding, so we should be safe for the foreseeable future ;)

Reply:
Good point.  Then would you mind updating this PR to use the -v2 version and we'll get it some more testing and review.